components/tikv_util/src/logger/mod.rs

```diff
@@ -116,11 +116,12 @@ pub type RotatingFileDecorator = PlainDecorator<BufWriter<RotatingFileLogger>>;
 /// Constructs a new file drainer which outputs log to a file at the specified
 /// path. The file drainer rotates for the specified timespan.
 pub fn file_drainer<N>(
+    format: LogFormat,
```

Comment:
How about removing the format argument and adding helper functions like `text_format` to construct the logger? So a file drainer that output text format becomes `text_format(file_drainer())`.

Reply:
Did you mean that let caller decide the format explicitly? So the logger module would provide two formatter: `text_format` and `json_format`. And then caller can compose different drainer and formatter at will.

But I encountered two problems. First, `fn json_format` now returns a `slog_json::Json` struct. And to consturct `slog_json::Json` we need a type impl `Write` trait, which is not compatible with current signature of `fn TikvFormat::new()`. Second, supposed `json_format` and `text_format` still return different types of `Drain`, and that would lead callers to handle dynamic dispatches themselves. The code may look like this:

```rust
let slow_log_drainer = logger::file_drainer(...)
let slow_log_drainer: Box<dyn slog::Drain<Ok = (), Err = io::Error> + Send> = match config.log_format {
    LogFormat::Text => Box::new(logger::text_format(slow_log_drainer)), // TikvFormat<_>
    LogFormat::Json => Box::new(logger::json_format(slow_log_drainer)), // slog_json::Json<_>
};
let drainer = logger::LogDispatcher::new(drainer, slow_log_drainer);
logger::init_log(
    drainer,
    // ...
)
```

I think this is not an ideal solution. Is there any better way to deal with it?

Reply:
You can change it to
```
let slow_log_drainer = logger::file_drainer(...)
match config.log_format {
    LogFormat::Text => build_logger(logger::text_format(slow_log_drainer)), // TikvFormat<_>
    LogFormat::Json => build_logger(logger::json_format(slow_log_drainer)), // slog_json::Json<_>
};

fn build_logger(slow_logger: impl Drain) {
    let drainer = logger::LogDispatcher::new(drainer, slow_log_drainer);
    logger::init_log(
        drainer,
        // ...
    )
}
```

Reply:
Implemented at a1acff1 and 2f69708. I've done followings changes but these are not elegant IMO:

1. To unify the signatures of `fn text_format` and `json_format`, `fn file_drainer(...) -> Drain` changes to `fn file_writer(...) -> Writer` in order to confront the generics type constraint.
2. in `setup.rs`, two local function `build_logger` and `build_loggers` are added to avoid additional `Box`es. However, these impls still need declare generic constraints explicitly and that may make code harder to grasp.

Let me know if you have any opinions. Thanks.